fractional-NFT/project/token-distributor-contract/src/main.sw

```diff
@@ -0,0 +1,240 @@
+        let mut token_distribution = require_fractional_nft_exists(storage.token_distributions.get(fractional_nft));
+
+        require(token_distribution.owner.is_some() && token_distribution.owner.unwrap() == msg_sender().unwrap(), AccessError::NotFNftOwner);
+        require(token_distribution.state == DistributionState::Distributed, AccessError::InvalidState);
+
+        // Ensure the seller is only providing the amount to buyback the number of tokens that have been sold
+        let token_supply = fractional_nft_supply(fractional_nft);
+        require(msg_amount() == (token_supply - this_balance(fractional_nft)) * token_price && msg_asset_id() == token_distribution.external_asset, AssetError::InvalidAssetTransfer);
+
+        // Store the buyback price and change the state
+        token_distribution.token_price = token_price;
+        token_distribution.state = DistributionState::Buyback;
+        storage.token_distributions.insert(fractional_nft, Option::Some(token_distribution));
+
+        log(Buyback {
+            fractional_nft,
+            token_price,
+        });
+    }
+
+    #[storage(read, write)]
+    fn create(
+        external_asset: ContractId,
+        fractional_nft: ContractId,
+        nft: ContractId,
+        owner: Option<Identity>,
+        reserve_price: Option<u64>,
+        token_price: u64,
+        token_supply: u64,
+        nft_token_id: u64,
+    ) {
+        require(storage.token_distributions.get(fractional_nft).is_none(), AccessError::DistributionAlreadyExists);
+
+        create_fractional_nft(fractional_nft, nft, Identity::ContractId(contract_id()), token_supply, nft_token_id);
+
+        // Store the newly created token distribution information
+        let token_distribution = TokenDistribution::new(external_asset, nft, owner, reserve_price, nft_token_id, token_price);
+        storage.token_distributions.insert(fractional_nft, Option::Some(token_distribution));
+
+        log(Created {
+            fractional_nft,
+            token_distribution,
+        });
+    }
+
+    #[storage(read, write)]
+    fn end(fractional_nft: ContractId) {
+        let mut token_distribution = require_fractional_nft_exists(storage.token_distributions.get(fractional_nft));
+
+        require(token_distribution.owner.is_some() && token_distribution.owner.unwrap() == msg_sender().unwrap(), AccessError::NotFNftOwner);
+        require(token_distribution.state == DistributionState::Buyback || token_distribution.state == DistributionState::Started, AccessError::InvalidState);
+
+        // Update distribution state
+        token_distribution.state = DistributionState::Ended;
+        storage.token_distributions.insert(fractional_nft, Option::Some(token_distribution));
+
+        // Send the remaining tokens to the fractionalized NFT contract and transfer ownership of the NFT to the owner
+        transfer(this_balance(fractional_nft), fractional_nft, Identity::ContractId(fractional_nft));
```

Comment:
We'd probably want the receiving contract to burn them but I'm unsure if we can or should enforce that somehow

Reply:
They are currently not burned by locked. I also don't believe it is possible to enforce burning